src/ModelContextProtocol/Shared/McpJsonRpcEndpoint.cs

```diff
@@ -122,6 +173,11 @@ public virtual async ValueTask DisposeUnsynchronizedAsync()
         _logger.EndpointCleanedUp(EndpointName);
     }
 
-    protected McpSession GetSessionOrThrow()
+    /// <summary>
+    /// Gets the current session.
+    /// </summary>
+    /// <returns>The current session.</returns>
+    /// <exception cref="InvalidOperationException">Thrown if the session is not started.</exception>
+    protected IMcpSession GetSessionOrThrow()
```

Comment:
```suggestion
    protected McpSession GetSessionOrThrow()
```

Even once we consolidate IMcpEnpoint and IMcpSession into a single, two-method interface, it doesn't makes sense to return the interface from here. Remember that only internal sealed types derive from this, so we might as well keep access to the fully fidelity type if we can.

Reply:
This is actually the part that breaks for me. Because McpJsonRpcEndpoint was internal, I changed it to public to make it testable - which I hate doing. I didn't want to go down the rabbit-hole of exposing McpSession and RequestHandlers just to be able to test a type.

Originally, I had added InternalsVisibleTo and made the test assembly a friend assembly. However, it requires a public key since the build signs the assembly. I went down a separate rabbit hole to only sign the assembly when the build configuration isn't set to DEBUG, but then ```make build``` and ```make test``` both fail, since they use release.

If there is a separate suggestion, I'm all for it. Not sure of a good pattern here for exposing the McpJsonRpcEndpoint type for testing...

Reply:
Don't use InternalsVisibleTo. We should not need to mock McpJsonRpcEndpoint. We should mock the transports or other dependencies for unit testing, but we shouldn't need to override internal sealed types. When I write tests for Kestrel, I don't override methods on the core [HttpProtocol](https://source.dot.net/#Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/HttpProtocol.cs,24be1f2a9d8270c0,references) abstract base type, that would be super fragile.

What exactly are you trying to test that you couldn't test by mocking transports and configuring the appropriate handlers?

Reply:
Since this type is the base type for a server and a client, I either need to duplicate my tests across all derived implementations, or test the base class. I'm really just testing that "SendRequestAsync" does the cancellation logic I expect. Ideally, I would want this done through composition instead of inheritance and make the cancellation logic optional / configurable for custom reason messages. I figured that'd be too large in scope for this changeset though. I'm trying to keep things small and clean.

Reply:
I'll admit internal access for unit tests can be nice. I still wouldn't change this to return an IMcpSession, but I might write tests against McpSession directly if it were public or we had internal visibility in our test. I'm looking into if we get get InternalsVisibleTo working at some point. Maybe by signing the test assembly.

In the meantime, I don't think it's too onerous to add two test cases to something like McpClientExtensionsTests. One each for the IMcpServer and IMcpClient. In McpClientExtensionsTests, you get an IMcpServer in the constructor that is the other side of the IMcpClient returned by CreateMcpClientForServer, so you can test both pretty easily using the same test class.

Reply:
(I don't want us to add InternalsVisibleTo.)

I'm not understanding what the concerns are here? Why isn't everything testable via the public surface area?